### ISSUE TITLE
"Equality [...] not allowed"

### ISSUE THREAD

**Issue by ArmborstL** [viperproject/silicon#484]
On the following code, I get the error _java.lang.AssertionError: assertion failed: Equality '(First:(First:($t@4@12)), Second:(First:($t@4@12))) == \_' is not allowed"_. It is not reported for a specific line of the code (_[1,1]_ instead), and changing different parts of the code makes the error disappear (e.g. removing the ```unfold```/```fold``` pair; removing ```bool_val``` from the assertion; changing it to ```==false```; including only one branch in the ```treePerm```).
What does this error mean, and how can I prevent it?
Thanks for the help!

PS: I'm using Viper 2.2.2.

```
field left: Ref
field right: Ref

predicate treePerm(current: Ref) {
  current != null 
          ==> acc(current.left, write) 
              && acc(current.right, write) 
}

method insert(current: Ref, bool_val: Bool)
  requires acc(treePerm(current), write)
  requires bool_val == true
{
  unfold acc(treePerm(current), write)
  fold acc(treePerm(current), write)
  assert (bool_val == true
          || (unfolding acc(treePerm(current), write) in true )
          )
}
```

**Comment by mschwerhoff** [comment]
Resolved with commit 75411d2

**Comment by mschwerhoff** [comment]
Resolved with commit 75411d2
